### PR TITLE
Implement vmware_argument_spec for required params

### LINKED ISSUE #25696: vmware_guest: non-helpful error when omitting required parameter..
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
vmware_guest

##### ANSIBLE VERSION

```
ansible 2.3.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.5 (default, Nov  6 2016, 00:28:07) [GCC 4.8.5 20150623 (Red Hat 4.8.5-11)]

```

##### CONFIGURATION
Default config, nothing changed

##### OS / ENVIRONMENT
CentOS 7

##### SUMMARY
I got the following error when trying to use vmware_guest:

```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: AttributeError: 'NoneType' object has no attribute 'encode'... <truncated>
```

##### STEPS TO REPRODUCE
Use vmware_guest module and omit the hostname parameter for vcenter

```yaml
- name: Clone template and customise
    vmware_guest:
      username: Administrator@vsphere.local
      password: "{{ vcenter_password }}"
      validate_certs: no
      datacenter: "{{ dc_name }}"
      esxi_hostname: "{{ esxi_hostname }}"
      name: "{{ vm_name|upper }}"
      state: "{{ vm_state }}"
      template: "{{ vm_template }}"
      networks:
      - name: "{{ vlan_id }}"
        ip: "{{ vm_ipaddress }}"
        netmask: "{{ vm_netmask }}"
        gateway: "{{ vm_gateway }}"
        domain: "{{ vm_domain }}"
      customization:
        autologon: true
        password: "{{ vm_password }}"
        joindomain: true
        domainadmin: "{{ domain_admin }}"
        domainadminpassword: "{{ domain_password }}"
```

##### EXPECTED RESULTS
I expected an error message telling me that a mandatory parameter was missing..

##### ACTUAL RESULTS
Ambiguous error message 

```
{
    "_ansible_parsed": false,
    "exception": "Traceback (most recent call last):\n  File \"/tmp/ansible_t0MsFQ/ansible_module_vmware_guest.py\", line 1332, in <module>\n    main()\n  File \"/tmp/ansible_t0MsFQ/ansible_module_vmware_guest.py\", line 1291, in main\n    pyv = PyVmomiHelper(module)\n  File \"/tmp/ansible_t0MsFQ/ansible_module_vmware_guest.py\", line 436, in __init__\n    self.content = connect_to_api(self.module)\n  File \"/tmp/ansible_t0MsFQ/ansible_modlib.zip/ansible/module_utils/vmware.py\", line 333, in connect_to_api\n  File \"/var/lib/awx/venv/ansible/lib/python2.7/site-packages/pyVim/connect.py\", line 817, in SmartConnect\n    sslContext)\n  File \"/var/lib/awx/venv/ansible/lib/python2.7/site-packages/pyVim/connect.py\", line 702, in __FindSupportedVersion\n    sslContext)\n  File \"/var/lib/awx/venv/ansible/lib/python2.7/site-packages/pyVim/connect.py\", line 622, in __GetServiceVersionDescription\n    path + \"/vimServiceVersions.xml\", sslContext)\n  File \"/var/lib/awx/venv/ansible/lib/python2.7/site-packages/pyVim/connect.py\", line 588, in __GetElementTree\n    conn.request(\"GET\", path)\n  File \"/usr/lib64/python2.7/httplib.py\", line 1017, in request\n    self._send_request(method, url, body, headers)\n  File \"/usr/lib64/python2.7/httplib.py\", line 1045, in _send_request\n    self.putrequest(method, url, **skips)\n  File \"/usr/lib64/python2.7/httplib.py\", line 946, in putrequest\n    host_enc = self.host.encode(\"ascii\")\nAttributeError: 'NoneType' object has no attribute 'encode'\n",
    "_ansible_no_log": false,
    "_ansible_delegated_vars": {
        "ansible_host": "localhost"
    },
    "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_t0MsFQ/ansible_module_vmware_guest.py\", line 1332, in <module>\n    main()\n  File \"/tmp/ansible_t0MsFQ/ansible_module_vmware_guest.py\", line 1291, in main\n    pyv = PyVmomiHelper(module)\n  File \"/tmp/ansible_t0MsFQ/ansible_module_vmware_guest.py\", line 436, in __init__\n    self.content = connect_to_api(self.module)\n  File \"/tmp/ansible_t0MsFQ/ansible_modlib.zip/ansible/module_utils/vmware.py\", line 333, in connect_to_api\n  File \"/var/lib/awx/venv/ansible/lib/python2.7/site-packages/pyVim/connect.py\", line 817, in SmartConnect\n    sslContext)\n  File \"/var/lib/awx/venv/ansible/lib/python2.7/site-packages/pyVim/connect.py\", line 702, in __FindSupportedVersion\n    sslContext)\n  File \"/var/lib/awx/venv/ansible/lib/python2.7/site-packages/pyVim/connect.py\", line 622, in __GetServiceVersionDescription\n    path + \"/vimServiceVersions.xml\", sslContext)\n  File \"/var/lib/awx/venv/ansible/lib/python2.7/site-packages/pyVim/connect.py\", line 588, in __GetElementTree\n    conn.request(\"GET\", path)\n  File \"/usr/lib64/python2.7/httplib.py\", line 1017, in request\n    self._send_request(method, url, body, headers)\n  File \"/usr/lib64/python2.7/httplib.py\", line 1045, in _send_request\n    self.putrequest(method, url, **skips)\n  File \"/usr/lib64/python2.7/httplib.py\", line 946, in putrequest\n    host_enc = self.host.encode(\"ascii\")\nAttributeError: 'NoneType' object has no attribute 'encode'\n",
    "changed": false,
    "module_stdout": "",
    "failed": true,
    "rc": 0,
    "msg": "MODULE FAILURE"
}
```

### PR DESCRIPTION
**Pull request by Akasurde**
##### SUMMARY
Without the fix hostname, username and password params
used to skip required check.

Fixes #25696

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
lib/ansible/modules/cloud/vmware/vmware_guest.py

##### ANSIBLE VERSION
```
2.4 devel
```

### CODE REVIEW

**Comment by Akasurde** [comment]
ready_for_review

File: lib/ansible/modules/cloud/vmware/vmware_guest_find.py [hunk 1]
```diff
@@ -239,27 +240,14 @@ def get_obj(content, vimtype, name):
...
-            ),
-            validate_certs=dict(required=False, type='bool', default=True),
-            name=dict(required=True, type='str'),
-            uuid=dict(required=False, type='str'),
-            datacenter=dict(required=True, type='str'),
-        ),
+    argument_spec = vmware_argument_spec()
+    argument_spec.update(
+        name=dict(required=True, type='str'),
+        uuid=dict(required=False, type='str'),
```
Comment (dagwieers) [inline lib/ansible/modules/cloud/vmware/vmware_guest_find.py]: Please do not add  `required=False`, it is implicit.

File: lib/ansible/modules/cloud/vmware/vmware_guest_facts.py [hunk 2]
```diff
@@ -167,29 +167,15 @@ def get_obj(content, vimtype, name):
...
-            name=dict(required=True, type='str'),
-            name_match=dict(required=False, type='str', default='first'),
-            uuid=dict(required=False, type='str'),
-            folder=dict(required=False, type='str', default='/vm'),
-            datacenter=dict(required=True, type='str'),
-        ),
+    argument_spec = vmware_argument_spec()
+    argument_spec.update(
+        name=dict(required=True, type='str'),
+        name_match=dict(required=False, type='str', default='first'),
```
Comment (dagwieers) [inline lib/ansible/modules/cloud/vmware/vmware_guest_facts.py]: Please remove `required=False`.

File: lib/ansible/modules/cloud/vmware/vmware_guest_facts.py [hunk 3]
```diff
@@ -167,29 +167,15 @@ def get_obj(content, vimtype, name):
...
+        uuid=dict(required=False, type='str'),
```
Comment (dagwieers) [inline lib/ansible/modules/cloud/vmware/vmware_guest_facts.py]: Please remove `required=False`.

File: lib/ansible/modules/cloud/vmware/vmware_guest_facts.py [hunk 4]
```diff
@@ -167,29 +167,15 @@ def get_obj(content, vimtype, name):
...
+        folder=dict(required=False, type='str', default='/vm'),
```
Comment (dagwieers) [inline lib/ansible/modules/cloud/vmware/vmware_guest_facts.py]: Please remove `required=False`.

File: lib/ansible/modules/cloud/vmware/vmware_guest_snapshot.py [hunk 5]
```diff
@@ -257,34 +257,21 @@ def apply_snapshot_op(self, vm):
...
-            uuid=dict(required=False, type='str'),
-            folder=dict(required=False, type='str', default='/vm'),
-            datacenter=dict(required=True, type='str'),
-            snapshot_name=dict(required=False, type='str'),
-            description=dict(required=False, type='str', default=''),
-        ),
+    argument_spec = vmware_argument_spec()
+    argument_spec.update(
+        state=dict(
+            required=False,
```
Comment (dagwieers) [inline lib/ansible/modules/cloud/vmware/vmware_guest_snapshot.py]: Please remove `required=False`.

File: lib/ansible/modules/cloud/vmware/vmware_guest_snapshot.py [hunk 6]
```diff
@@ -257,34 +257,21 @@ def apply_snapshot_op(self, vm):
...
+            choices=['present', 'absent', 'revert', 'remove_all'],
+            default='present'),
+        name=dict(required=True, type='str'),
+        name_match=dict(required=False, type='str', default='first'),
```
Comment (dagwieers) [inline lib/ansible/modules/cloud/vmware/vmware_guest_snapshot.py]: Please remove `required=False`.

File: lib/ansible/modules/cloud/vmware/vmware_guest_snapshot.py [hunk 7]
```diff
@@ -257,34 +257,21 @@ def apply_snapshot_op(self, vm):
...
+        uuid=dict(required=False, type='str'),
```
Comment (dagwieers) [inline lib/ansible/modules/cloud/vmware/vmware_guest_snapshot.py]: Please remove `required=False`.

File: lib/ansible/modules/cloud/vmware/vmware_guest_snapshot.py [hunk 8]
```diff
@@ -257,34 +257,21 @@ def apply_snapshot_op(self, vm):
...
+        folder=dict(required=False, type='str', default='/vm'),
```
Comment (dagwieers) [inline lib/ansible/modules/cloud/vmware/vmware_guest_snapshot.py]: Please remove `required=False`.

File: lib/ansible/modules/cloud/vmware/vmware_guest_snapshot.py [hunk 9]
```diff
@@ -257,34 +257,21 @@ def apply_snapshot_op(self, vm):
...
+        datacenter=dict(required=True, type='str'),
+        snapshot_name=dict(required=False, type='str'),
```
Comment (dagwieers) [inline lib/ansible/modules/cloud/vmware/vmware_guest_snapshot.py]: Please remove `required=False`.

File: lib/ansible/modules/cloud/vmware/vmware_guest_snapshot.py [hunk 10]
```diff
@@ -257,34 +257,21 @@ def apply_snapshot_op(self, vm):
...
+        description=dict(required=False, type='str', default=''),
```
Comment (dagwieers) [inline lib/ansible/modules/cloud/vmware/vmware_guest_snapshot.py]: Please remove `required=False`.

**Review by dagwieers — CHANGES_REQUESTED** [review]
For parameters I prefer to add options that are prevalent, over options that are optional, or have implicit values, or are seldom added. So I prefer to use the following order: `type`, `required`, `default`, `choices`, `aliases`, `no_log` . (`required` is a second, because it is also more important and mutual exclusive with `default`)

**Comment by Akasurde** [comment]
@dagwieers Done.

**Comment by Akasurde** [comment]
ready_for_review

**Comment by mattclay** [comment]
CI failure in integration test `test/integration/targets/vmware_guest_find/tasks/main.yml:35 / [testhost] testhost: vmware_guest_find : find folders for each vm username={{ vcsim_instance['json']['username'] }}, datacenter={{ (item|basename).split('_')[0] }}, hostname={{ vcsim }}, name={{ item|basename }}, password={{ vcsim_instance['json']['password'] }}, validate_certs=False`:

```
{
"changed": false, 
"failed": true, 
"msg": "All items completed"
}
```

**Comment by Akasurde** [comment]
Failures in last CI run are not related to the change. Closing and re-opening for CI.

**Comment by Akasurde** [comment]
ready_for_review

**Comment by jctanner** [comment]
@Akasurde this needs a rebase after the previous merge on vmware_guest_snapshot

**Comment by Akasurde** [comment]
@jctanner ready_for_review

**Comment by Akasurde** [comment]
@jctanner @dagwieers @mattclay Thanks for review and comments.

**Pull request merged**